sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/JavaOrcColumnarBatchReader.java

```diff
@@ -0,0 +1,503 @@
+      vector.reset();
+    }
+    columnarBatch.setNumRows(batchSize);
+    int i = 0;
+    while (i < requiredSchema.length()) {
+      StructField field = requiredSchema.fields()[i];
+      WritableColumnVector toColumn = columnVectors[i];
+
+      if (requestedColIds[i] < 0) {
+        toColumn.appendNulls(batchSize);
+      } else {
+        ColumnVector fromColumn = batch.cols[requestedColIds[i]];
+
+        if (fromColumn.isRepeating) {
+          if (fromColumn.isNull[0]) {
+            toColumn.appendNulls(batchSize);
+          } else {
+            DataType type = field.dataType();
+            if (type instanceof BooleanType) {
+              toColumn.appendBooleans(batchSize, ((LongColumnVector)fromColumn).vector[0] == 1);
+            } else if (type instanceof ByteType) {
+              toColumn.appendBytes(batchSize, (byte)((LongColumnVector)fromColumn).vector[0]);
+            } else if (type instanceof ShortType) {
+              toColumn.appendShorts(batchSize, (short)((LongColumnVector)fromColumn).vector[0]);
+            } else if (type instanceof IntegerType || type instanceof DateType) {
+              toColumn.appendInts(batchSize, (int)((LongColumnVector)fromColumn).vector[0]);
+            } else if (type instanceof LongType) {
+              toColumn.appendLongs(batchSize, ((LongColumnVector)fromColumn).vector[0]);
+            } else if (type instanceof TimestampType) {
+              toColumn.appendLongs(batchSize, fromTimestampColumnVector((TimestampColumnVector)fromColumn, 0));
+            } else if (type instanceof FloatType) {
+              toColumn.appendFloats(batchSize, (float)((DoubleColumnVector)fromColumn).vector[0]);
+            } else if (type instanceof DoubleType) {
+              toColumn.appendDoubles(batchSize, ((DoubleColumnVector)fromColumn).vector[0]);
+            } else if (type instanceof StringType || type instanceof BinaryType) {
+              BytesColumnVector data = (BytesColumnVector)fromColumn;
+              int index = 0;
+              while (index < batchSize) {
+                toColumn.appendByteArray(data.vector[0], data.start[0], data.length[0]);
+                index += 1;
+              }
+            } else if (type instanceof DecimalType) {
+              DecimalType decimalType = (DecimalType)type;
+              appendDecimalWritable(
+                toColumn,
+                decimalType.precision(),
+                decimalType.scale(),
+                ((DecimalColumnVector)fromColumn).vector[0]);
+            } else {
+              throw new UnsupportedOperationException("Unsupported Data Type: " + type);
+            }
+          }
+        } else if (fromColumn.noNulls) {
+          DataType type = field.dataType();
+          if (type instanceof BooleanType) {
+            long[] data = ((LongColumnVector)fromColumn).vector;
+            int index = 0;
+            while (index < batchSize) {
+              toColumn.appendBoolean(data[index] == 1);
```

Comment:
Actually, you know the number of rows in advance. Wouldn't it possible to call reserve() once, and then use the putX() API instead of appendX() API inside the loops? That should be significantly faster.

Reply:
good catch! Since we allocated a big enough batch in `initBatch`, I think here we should call the `putXXX` API.

Reply:
I checked this bit more, and it seems that the code path after putX() is bimorphic and 100% biased. If you split each for loop to it's own method and use putX(),  then the entire contents of the for loops will be basically reduced to a single call of Unsafe.putX().

Reply:
Thank you, @henrify . I'll update like that.

Reply:
Thinking this further, the ColumnVector API supports direct memory copy. For long, double, binary and string you might be able to scrap the for loops entirely, and copy ORC's memory directly to Spark's memory, with just 1 call per batch, which would be way faster than looping. Check e.g. putDoubles(int rowId, int count, double[] src, int srcIndex).

Reply:
For long and double, we don't have loop already.
For `binary` and `string`, is it possible?

Reply:
Ah that's right. I read the code for the nullable variant, which uses (and needs to use) loop.